File: windows/src/RapiDrop.Core/Security/CryptoEngine.cs
using System.Security.Cryptography;
using System.Text;

namespace RapiDrop.Core.Security;

public record HandshakeKeys(
    byte[] SessionKey,
    byte[] AuthKey,
    byte[] PairRecordKey,
    string SasCode,
    byte[] TranscriptHash
);

public static class CryptoEngine
{
    private const int NonceSize = 12;
    private const int TagSize = 16;
    private const int KeySize = 32;

    public static byte[] DeriveKeyFromPin(string pin)
    {
        string cleanPin = new string(pin.Where(char.IsDigit).ToArray());
        byte[] secret = Encoding.UTF8.GetBytes($"RapiDrop-PIN-{cleanPin}");
        return SHA256.HashData(secret);
    }

    public static (byte[] PrivateKey, string PublicKeyHex) GenerateEphemeralKeypair()
    {
        byte[] priv = Curve25519.GeneratePrivateKey();
        byte[] pub = Curve25519.ComputePublicKey(priv);
        return (priv, Convert.ToHexString(pub).ToLowerInvariant());
    }

    public static byte[] GenerateNonce(int count = 16)
    {
        byte[] nonce = new byte[count];
        RandomNumberGenerator.Fill(nonce);
        return nonce;
    }

    public static byte[]? HexToBytes(string hex)
    {
        string clean = hex.Trim();
        if (clean.Length % 2 != 0) return null;
        try
        {
            return Convert.FromHexString(clean);
        }
        catch
        {
            return null;
        }
    }

    public static byte[] ComputeSharedSecret(byte[] privateKey, string remotePublicKeyHex)
    {
        byte[]? remotePub = HexToBytes(remotePublicKeyHex);
        if (remotePub == null || remotePub.Length != 32)
            throw new ArgumentException("Invalid public key hex");
        return Curve25519.ComputeSharedSecret(privateKey, remotePub);
    }

    public static HandshakeKeys DeriveHandshakeKeys(
        byte[] sharedSecret,
        byte[] initiatorPublicKey,
        byte[] receiverPublicKey,
        byte[] initiatorNonce,
        byte[] receiverNonce,
        string initiatorId,
        string receiverId
    )
    {
        byte[] prefix = Encoding.UTF8.GetBytes("RapiDrop-v1|");
        byte[] pipe = Encoding.UTF8.GetBytes("|");
        byte[] idA = Encoding.UTF8.GetBytes(initiatorId);
        byte[] idB = Encoding.UTF8.GetBytes(receiverId);

        using var ms = new MemoryStream();
        ms.Write(prefix);
        ms.Write(initiatorPublicKey);
        ms.Write(pipe);
        ms.Write(receiverPublicKey);
        ms.Write(pipe);
        ms.Write(initiatorNonce);
        ms.Write(pipe);
        ms.Write(receiverNonce);
        ms.Write(pipe);
        ms.Write(idA);
        ms.Write(pipe);
        ms.Write(idB);
        byte[] transcript = ms.ToArray();
        byte[] transcriptHash = SHA256.HashData(transcript);

        byte[] salt = new byte[initiatorNonce.Length + receiverNonce.Length];
        Buffer.BlockCopy(initiatorNonce, 0, salt, 0, initiatorNonce.Length);
        Buffer.BlockCopy(receiverNonce, 0, salt, initiatorNonce.Length, receiverNonce.Length);

        byte[] prk = HKDF.Extract(HashAlgorithmName.SHA256, sharedSecret, salt);

        byte[] sessInfo = Encoding.UTF8.GetBytes("RapiDrop-v1-Session-Key|").Concat(transcriptHash).ToArray();
        byte[] sessionKey = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32, sessInfo);

        byte[] authInfo = Encoding.UTF8.GetBytes("RapiDrop-v1-Auth-Token|").Concat(transcriptHash).ToArray();
        byte[] authKey = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32, authInfo);

        byte[] pairInfo = Encoding.UTF8.GetBytes("RapiDrop-v1-Pair-Record|").Concat(transcriptHash).ToArray();
        byte[] pairRecordKey = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32, pairInfo);

        uint val32 = ((uint)authKey[0] << 24) |
                     ((uint)authKey[1] << 16) |
                     ((uint)authKey[2] << 8) |
                     (uint)authKey[3];
        uint sasNum = val32 % 1_000_000;
        string sasCode = sasNum.ToString("D6");

        return new HandshakeKeys(sessionKey, authKey, pairRecordKey, sasCode, transcriptHash);
    }

    public static (byte[] Ciphertext, byte[] Nonce, byte[] Tag) Encrypt(byte[] payload, byte[] key)
    {
        if (key.Length != KeySize) throw new ArgumentException($"Key must be {KeySize} bytes", nameof(key));

        byte[] nonce = new byte[NonceSize];
        RandomNumberGenerator.Fill(nonce);

        byte[] ciphertext = new byte[payload.Length];
        byte[] tag = new byte[TagSize];

        using var aesGcm = new AesGcm(key, TagSize);
        aesGcm.Encrypt(nonce, payload, ciphertext, tag);

        return (ciphertext, nonce, tag);
    }

    public static byte[] Decrypt(byte[] ciphertext, byte[] nonce, byte[] tag, byte[] key)
    {
        if (key.Length != KeySize) throw new ArgumentException($"Key must be {KeySize} bytes", nameof(key));
        if (nonce.Length != NonceSize) throw new ArgumentException($"Nonce must be {NonceSize} bytes", nameof(nonce));
        if (tag.Length != TagSize) throw new ArgumentException($"Tag must be {TagSize} bytes", nameof(tag));

        byte[] plaintext = new byte[ciphertext.Length];

        using var aesGcm = new AesGcm(key, TagSize);
        aesGcm.Decrypt(nonce, ciphertext, tag, plaintext);

        return plaintext;
    }

    public static bool TryDecrypt(byte[] ciphertext, byte[] nonce, byte[] tag, byte[] key, out byte[]? plaintext)
    {
        try
        {
            plaintext = Decrypt(ciphertext, nonce, tag, key);
            return true;
        }
        catch
        {
            plaintext = null;
            return false;
        }
    }
}

File: android/app/src/main/java/com/prabotics/rapidrop/security/CryptoEngine.kt
package com.prabotics.rapidrop.security

import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec



data class EncryptionResult(
    val ciphertext: ByteArray,
    val nonce: ByteArray,
    val tag: ByteArray
)

data class HandshakeKeys(
    val sessionKey: SecretKey,
    val authKey: ByteArray,
    val pairRecordKey: ByteArray,
    val sasCode: String,
    val transcriptHash: ByteArray
)

object CryptoEngine {
    private const val GCM_TAG_LENGTH_BITS = 128
    private const val GCM_NONCE_LENGTH = 12
    private val secureRandom = SecureRandom()
    private val cipherThreadLocal = ThreadLocal.withInitial {
        Cipher.getInstance("AES/GCM/NoPadding")
    }

    private fun getCipher(): Cipher =
        cipherThreadLocal.get() ?: Cipher.getInstance("AES/GCM/NoPadding")

    fun deriveKeyFromPin(pin: String): SecretKey {
        val cleanPin = pin.filter { it.isDigit() }
        val secret = "RapiDrop-PIN-$cleanPin".toByteArray(Charsets.UTF_8)
        val md = MessageDigest.getInstance("SHA-256")
        val keyBytes = md.digest(secret)
        return SecretKeySpec(keyBytes, "AES")
    }

    fun generateEphemeralKeypair(): Pair<ByteArray, String> {
        val priv = Curve25519.generatePrivateKey(secureRandom)
        val pub = Curve25519.computePublicKey(priv)
        return Pair(priv, bytesToHex(pub))
    }

    fun generateNonce(count: Int = 16): ByteArray {
        val nonce = ByteArray(count)
        secureRandom.nextBytes(nonce)
        return nonce
    }

    fun bytesToHex(bytes: ByteArray): String {
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            sb.append(String.format("%02x", b.toInt() and 0xFF))
        }
        return sb.toString()
    }

    fun hexToBytes(hex: String): ByteArray? {
        val clean = hex.trim()
        if (clean.length % 2 != 0) return null
        val result = ByteArray(clean.length / 2)
        for (i in result.indices) {
            val byteStr = clean.substring(i * 2, i * 2 + 2)
            result[i] = byteStr.toIntOrNull(16)?.toByte() ?: return null
        }
        return result
    }

    fun computeSharedSecret(privateKey: ByteArray, remotePublicKeyHex: String): ByteArray {
        val remotePub = hexToBytes(remotePublicKeyHex) ?: throw IllegalArgumentException("Invalid public key hex")
        if (remotePub.size != 32) throw IllegalArgumentException("Invalid public key size")
        return Curve25519.computeSharedSecret(privateKey, remotePub)
    }

    fun deriveHandshakeKeys(
        sharedSecret: ByteArray,
        initiatorPublicKey: ByteArray,
        receiverPublicKey: ByteArray,
        initiatorNonce: ByteArray,
        receiverNonce: ByteArray,
        initiatorId: String,
        receiverId: String
    ): HandshakeKeys {
        val pipe = "|".toByteArray(Charsets.UTF_8)
        val transcript = ByteArray(
            "RapiDrop-v1|".toByteArray(Charsets.UTF_8).size +
            initiatorPublicKey.size + pipe.size +
            receiverPublicKey.size + pipe.size +
            initiatorNonce.size + pipe.size +
            receiverNonce.size + pipe.size +
            initiatorId.toByteArray(Charsets.UTF_8).size + pipe.size +
            receiverId.toByteArray(Charsets.UTF_8).size
        )

        var offset = 0
        fun append(bytes: ByteArray) {
            System.arraycopy(bytes, 0, transcript, offset, bytes.size)
            offset += bytes.size
        }

        append("RapiDrop-v1|".toByteArray(Charsets.UTF_8))
        append(initiatorPublicKey)
        append(pipe)
        append(receiverPublicKey)
        append(pipe)
        append(initiatorNonce)
        append(pipe)
        append(receiverNonce)
        append(pipe)
        append(initiatorId.toByteArray(Charsets.UTF_8))
        append(pipe)
        append(receiverId.toByteArray(Charsets.UTF_8))
        val md = MessageDigest.getInstance("SHA-256")
        val transcriptHash = md.digest(transcript)

        val salt = ByteArray(initiatorNonce.size + receiverNonce.size)
        System.arraycopy(initiatorNonce, 0, salt, 0, initiatorNonce.size)
        System.arraycopy(receiverNonce, 0, salt, initiatorNonce.size, receiverNonce.size)

        val prk = hkdfExtract(salt, sharedSecret)

        val sessInfo = "RapiDrop-v1-Session-Key|".toByteArray(Charsets.UTF_8) + transcriptHash
        val sessionKeyBytes = hkdfExpand(prk, sessInfo, 32)
        val sessionKey = SecretKeySpec(sessionKeyBytes, "AES")

        val authInfo = "RapiDrop-v1-Auth-Token|".toByteArray(Charsets.UTF_8) + transcriptHash
        val authKey = hkdfExpand(prk, authInfo, 32)

        val pairInfo = "RapiDrop-v1-Pair-Record|".toByteArray(Charsets.UTF_8) + transcriptHash
        val pairRecordKey = hkdfExpand(prk, pairInfo, 32)

        val val32 = ((authKey[0].toInt() and 0xFF).toLong() shl 24) or
                    ((authKey[1].toInt() and 0xFF).toLong() shl 16) or
                    ((authKey[2].toInt() and 0xFF).toLong() shl 8) or
                    (authKey[3].toInt() and 0xFF).toLong()
        val sasNum = (val32 % 1_000_000).toInt()
        val sasCode = String.format(java.util.Locale.US, "%06d", sasNum)

        return HandshakeKeys(
            sessionKey = sessionKey,
            authKey = authKey,
            pairRecordKey = pairRecordKey,
            sasCode = sasCode,
            transcriptHash = transcriptHash
        )
    }

    private fun hkdfExtract(salt: ByteArray, ikm: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        val actualSalt = if (salt.isNotEmpty()) salt else ByteArray(32)
        mac.init(SecretKeySpec(actualSalt, "HmacSHA256"))
        return mac.doFinal(ikm)
    }

    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val okm = ByteArray(length)
        var t = ByteArray(0)
        var offset = 0
        var i = 1
        while (offset < length) {
            mac.reset()
            mac.update(t)
            mac.update(info)
            mac.update(i.toByte())
            t = mac.doFinal()
            val toCopy = Math.min(t.size, length - offset)
            System.arraycopy(t, 0, okm, offset, toCopy)
            offset += toCopy
            i++
        }
        return okm
    }

    fun encrypt(payload: ByteArray, key: SecretKey): EncryptionResult =
        encrypt(payload, 0, payload.size, key)

    fun encrypt(input: ByteArray, offset: Int, length: Int, key: SecretKey): EncryptionResult {
        val nonce = ByteArray(GCM_NONCE_LENGTH)
        secureRandom.nextBytes(nonce)

        val cipher = getCipher()
        val spec = GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce)
        cipher.init(Cipher.ENCRYPT_MODE, key, spec)
        val cipherWithTag = cipher.doFinal(input, offset, length)
        val tagLength = GCM_TAG_LENGTH_BITS / 8
        val ciphertextLength = cipherWithTag.size - tagLength

        val ciphertext = ByteArray(ciphertextLength)
        val tag = ByteArray(tagLength)

        System.arraycopy(cipherWithTag, 0, ciphertext, 0, ciphertextLength)
        System.arraycopy(cipherWithTag, ciphertextLength, tag, 0, tagLength)

        return EncryptionResult(ciphertext, nonce, tag)
    }

    fun decrypt(
        ciphertext: ByteArray,
        nonce: ByteArray,
        tag: ByteArray,
        key: SecretKey
    ): ByteArray {
        val cipher = getCipher()
        val spec = GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce)
        cipher.init(Cipher.DECRYPT_MODE, key, spec)
        val p1 = cipher.update(ciphertext)
        val p2 = cipher.doFinal(tag)
        return when {
            p1 == null || p1.isEmpty() -> p2 ?: ByteArray(0)
            p2 == null || p2.isEmpty() -> p1
            else -> {
                val combined = ByteArray(p1.size + p2.size)
                System.arraycopy(p1, 0, combined, 0, p1.size)
                System.arraycopy(p2, 0, combined, p1.size, p2.size)
                combined
            }
        }
    }
}
